port/unix/omrsysinfo.c

```diff
@@ -3970,6 +4024,77 @@ getHandleOfCgroupSubsystemFile(struct OMRPortLibrary *portLibrary, uint64_t subs
 	return rc;
 }
 
+/**
+ * Reads the Cgroup metric (if has multiple searched for metric key and reads it) and returns the value to the param `value`
+ *
+ * @param[in] portLibrary pointer to OMRPortLibrary
+ * @param[in] subsystemFlag flag of type OMR_CGROUP_SUBSYSTEMS_* representing the cgroup subsystem
+ * @param[in] fileName name of the file under cgroup subsystem
+ * @param[in] metricKeyInFile name of the cgroup metric which is in a group of metrics in a cgroup subsystem file
+ * @param[in/out] value pointer to the buffer where the value of metric is returned
+ * @param[in/out] fileContent double pointer to allocate memory and write file content in case of a multiline file with more than 1 metric available in it
+ * @param[in] isSingleLineKeyValue BOOLEAN to indicate that the file has a single metric in a key value representation
+ * @param[in] sizeRef An integer to check the buffer size of the value param by having this as reference
+ * @param[in/out] value pointer to char * which stores the value for the specified cgroup subsystem metric
+ *
+ * @return 0 on success, negative error code on any error
+ */
+static int32_t
+readCgroupMetricFromFile(struct OMRPortLibrary *portLibrary, uint64_t subsystemFlag, const char *fileName, const char *metricKeyInFile, char *value, char **fileContent, BOOLEAN isSingleLineKeyValue, uint64_t sizeRef)
+{
+	int32_t rc = 0;
+	FILE *file = NULL;
+	rc = getHandleOfCgroupSubsystemFile(portLibrary, subsystemFlag, fileName, &file);
+	if (0 != rc) {
+		goto _end;
+	}
+	if (NULL == metricKeyInFile) { /* If `metricKeyInFile` is NULL then it states the file has only one Cgroup metric to be read */
+		if (NULL == fgets(value, MAX_LINE_LENGTH, file)) {
+			rc = OMRPORT_ERROR_FILE_OPFAILED;
+		}
+		goto _end;
+	} else if (isSingleLineKeyValue) { /* Else if its a single lined entry with metric, we search for the metric and get its value */
+		char bufferStore[MAX_LINE_LENGTH] = {0};
+		size_t size = strlen(metricKeyInFile);
+		if (NULL != fgets(bufferStore, MAX_LINE_LENGTH, file)) {
+			char *tmpPtr = bufferStore;
+			if (0 == strncmp(tmpPtr, metricKeyInFile, size)) {
+				tmpPtr += size;
+				if (strlen(tmpPtr) <= sizeRef){
+					sscanf(tmpPtr, "%s", value);
+				} else {
+					rc = OMRPORT_ERROR_STRING_BUFFER_TOO_SMALL;
+				}
+				goto _end;
+			}
+		} else {
+			rc = OMRPORT_ERROR_FILE_OPFAILED;
+			goto _end;
+		}
+	} else { /* Else we read the file to a buffer and pick the metrics from it later */
+		*fileContent = portLibrary->mem_allocate_memory(portLibrary, 1024, OMR_GET_CALLSITE(), OMRMEM_CATEGORY_PORT_LIBRARY);
+		if (NULL != *fileContent) {
+			size_t returnvalue = fread(*fileContent, 1, 1024, file);
+			if (0 == returnvalue) {
+				portLibrary->mem_free_memory(portLibrary, *fileContent);
+				rc = OMRPORT_ERROR_FILE_OPFAILED;
+				goto _end;
+			}
+		} else {
+			portLibrary->mem_free_memory(portLibrary, *fileContent);
```

Comment:
The `else` case pointed out here was in the case of memory allocation failure as `*fileContent` will be `NULL`, i have made the code to return `OMRPORT_ERROR_SYSINFO_MEMORY_ALLOC_FAILED`. 

The pointer is free'd in case of file read failure and in end case of [all the metrics are read from it](https://github.com/eclipse/omr/pull/2994#discussion_r225845837) . So i'm thinking in any of the cases we are not letting the free'd pointer to be leaked to user. 

The caller will have access to it in the process of the iterating in between the successive calls to `omrsysinfo_cgroup_subsystem_iterator_next` as its the part of struct `OMRCgroupMetricIteratorState` which is been passed by the caller. At the end of the iterating every subsystem its free'd.

@DanHeidinga Please correct me if i'm wrong and let me know the changes i need to do accordingly.

Thanks